crates/package_manager/src/virtual_dir.rs

```diff
@@ -0,0 +1,79 @@
+use crate::{link_file, symlink_pkg, LinkFileError};
+use derive_more::{Display, Error};
+use miette::Diagnostic;
+use pacquet_lockfile::{
+    DependencyPath, PackageSnapshot, PackageSnapshotDependency, PkgNameVerPeer,
+};
+use pacquet_npmrc::PackageImportMethod;
+use rayon::prelude::*;
+use std::{
+    collections::HashMap,
+    ffi::OsString,
+    fs,
+    path::{Path, PathBuf},
+};
+
+#[derive(Debug, Display, Error, Diagnostic)]
+pub enum CreateVirtdirError {
+    #[diagnostic(transparent)]
+    LinkFile(#[error(source)] LinkFileError),
+}
+
+/// This function does 2 things:
+/// 1. Install the files from `cas_paths`
+/// 2. Create the symlink layout
+///
+/// **TODO:** may break this function into 2 later
+pub fn create_virtdir_by_snapshot(
+    dependency_path: &DependencyPath,
+    virtual_store_dir: &Path,
+    cas_paths: &HashMap<OsString, PathBuf>,
+    import_method: PackageImportMethod,
+    package_snapshot: &PackageSnapshot,
+) -> Result<(), CreateVirtdirError> {
+    assert_eq!(
+        import_method,
+        PackageImportMethod::Auto,
+        "Only auto import method is supported, but {dependency_path} requires {import_method:?}",
+    );
+
+    // node_modules/.pacquet/pkg-name@x.y.z/node_modules
+    let virtual_node_modules_dir = virtual_store_dir
+        .join(dependency_path.package_specifier.to_virtual_store_name())
+        .join("node_modules");
+    fs::create_dir_all(&virtual_node_modules_dir).unwrap_or_else(|error| {
+        panic!("Failed to create directory at {virtual_node_modules_dir:?}: {error}")
+    }); // TODO: proper error propagation
+
+    // 1. Install the files from `cas_paths`
+    let save_path =
+        virtual_node_modules_dir.join(dependency_path.package_specifier.name.to_string());
+    if !save_path.exists() {
+        cas_paths.par_iter().try_for_each(|(cleaned_entry, store_path)| {
+            link_file(store_path, &save_path.join(cleaned_entry))
+                .map_err(CreateVirtdirError::LinkFile)
+        })?;
+    }
+
+    // 2. Create the symlink layout
+    if let Some(dependencies) = &package_snapshot.dependencies {
+        dependencies.par_iter().for_each(|(name, spec)| {
+            let virtual_store_name = match spec {
+                PackageSnapshotDependency::PkgVerPeer(ver_peer) => {
+                    let package_specifier = PkgNameVerPeer::new(name.clone(), ver_peer.clone()); // TODO: remove copying here
```

Comment:
PkgNameVerPeer is really ambiguous. I PackageNameVersionPeer also doesn't make sense as a struct name.

Reply:
`PkgNameVerPeer` vs `PackageNameVersionPeer` preference aside. Naming this was particularly difficult for me, because I had to follow the lockfile format as defined in pnpm.

In pnpm, there are peculiar anonymous string types with defined syntax. The pnpm repo doesn't give them a name, only use `string` and parse it later (which is error-prone).

In Rust, I made it parses at the same them as `serde`. Though the names of these data structure is not clear as their purposes are overlapped and reused. Perhaps @zkochan can help us find better names for these `Pkg*` types?

Reply:
I am not sure. Dependency path could be a good name because this is basically the directory name inside the virtual store. But dep path is already used for this + the custom registry. Maybe ShortDependencyPath? Then the one currently called dependency path would become FullDependencyPath?

And by the way, a short dependency path can be only a registry-hosted package. githosted and other types of dependencies have a different format for the dependency path.

Reply:
> Dependency path could be a good name because this is basically the directory name inside the virtual store. 

It's not exactly the names in the virtual store. The names in the lockfile uses `(` and `)`, but they were all replaced with `_`. Furthermore, the names in the lockfile contain `/` for scoped names. I originally name it `DependencyPath` because the spec calls the keys of the snapshot object "dependency path".

TBH, I don't think this is a good name. It doesn't refer to any actual path, and it fails to describe the purpose of the type as well as the components.

Concatenating the components of a syntax together may have been a temporary solution at the time, but now that I think about it, it's the best solution so far.

> And by the way, a short dependency path can be only a registry-hosted package. githosted and other types of dependencies have a different format for the dependency path.

Does this mean that this "short dependency path" should be an enum and name+ver+peer is a variant in it?

Reply:
> It's not exactly the names in the virtual store. The names in the lockfile uses ( and ), but they were all replaced with _. Furthermore, the names in the lockfile contain / for scoped names.

Like you said, it is the directory name with some conversion to make it a valid directory name.

> Concatenating the components of a syntax together may have been a temporary solution at the time, but now that I think about it, it's the best solution so far.

You mean naming it PkgNameVerPeers? But then you also need to add PatchHash to it because it also can contain a patch hash.

> Does this mean that this "short dependency path" should be an enum and name+ver+peer is a variant in it?

no, because you separated the custom registry field from name+ver+peer and custom registry is only related to name+ver+peer, not other types of "dependency path"

Maybe it can be named "Dependency ID"? We currently use Package ID to name name+ver. So name+ver+peers would be Dependency ID. But it could also be PkgSnapshotID as the entries in the lockfile are called PackageSnapshot.

Reply:
@zkochan Let's move the discussion here: https://github.com/pnpm/pacquet/issues/146